dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1070 @@
+    size_type num_variables() const { return linear_biases_.size(); }
+
+    /// Return the offset
+    bias_type offset() const { return offset_; }
+
+    /**
+     * Return the quadratic bias associated with `u`, `v`.
+     *
+     * If `u` and `v` do not have a quadratic bias, returns 0.
+     *
+     * Note that this function does not return a reference, this is because
+     * each quadratic bias is stored twice.
+     *
+     */
+    bias_type quadratic(index_type u, index_type v) const {
+        if (!adj_ptr_) {
+            return 0;
+        }
+        return (*adj_ptr_)[u].get(v);
+    }
+
+    /**
+     * Return the quadratic bias associated with `u`, `v`.
+     *
+     * Note that this function does not return a reference, this is because
+     * each quadratic bias is stored twice.
+     *
+     * Raises an `out_of_range` error if either `u` or `v` are not variables or
+     * if they do not have an interaction then the function throws an exception.
+     */
+    bias_type quadratic_at(index_type u, index_type v) const {
+        if (!adj_ptr_) {
+            throw std::out_of_range("given variables have no interaction");
+        }
+        return (*adj_ptr_)[u].at(v);
+    }
+
+    /// Remove the interaction between variables `u` and `v`.
+    bool remove_interaction(index_type u, index_type v) {
+        if (has_adj() && (*adj_ptr_)[u].erase(v)) {
+            if (u != v) {
+                (*adj_ptr_)[v].erase(u);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Remove variable `v` from the model.
+     * 
+     * Note that this causes a reindexing, where all variables above `v` have
+     * their index reduced by one.
+     */
+    void remove_variable(index_type v) {
+        assert(0 <= v && static_cast<size_type>(v) < num_variables());
+
+        linear_biases_.erase(linear_biases_.cbegin() + v);
+
+        if (has_adj()) {
+            // remove v's neighborhood
+            adj_ptr_->erase(adj_ptr_->cbegin() + v);
+
+            for (auto& n : *adj_ptr_) {
+                // work backwards through each neighborhood, decrementing the indices above v
+                // we'll work with a regular iterator rather than a reverse one because
+                // we'll eventually call erase
+                for (auto it = n.end(); it != n.begin();) {
+                    --it;
+
+                    if (it->v > v) {
+                        // decrement the index
+                        --(it->v);
+                    } else if (it->v == v) {
+                        // remove this element and break
+                        n.erase(it);
+                        break;
+                    } else {
+                        // we're done
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    /// Multiply all biases 'scalar'
+    void scale(bias_type scalar) {
```

Comment:
You meant "scaler" or "scalar" ?

Reply:
[scalar](https://en.wikipedia.org/wiki/Scalar_(mathematics)) is the math term